### PR TITLE
Optimization of recommendations sorting

### PR DESCRIPTION
**Pull request by netang**
adding filter_sort function instead of udf sorter function for optimization purposes

### CODE REVIEW

File: replay/metrics/base_metric.py [hunk 1]
```diff
@@ -99,6 +99,23 @@ def preprocess_gt(
     return true_items_by_users
 
 
+def filter_sort(recommendations: AnyDataFrame) -> DataFrame:
```
Comment (monkey0head) [inline replay/metrics/base_metric.py:102]: you add new function `filter_sort` but do not remove the old one, `sorter`. where is it used for now (at least NCIS metric)? do we need to completely remove `sorter`? `sorter` allows to return an extra data. need to discuss

File: tests/test_metrics.py [hunk 2]
```diff
@@ -423,17 +423,6 @@ def test_duplicate_recs(quality_metrics, duplicate_recs, recs, true):
         )
 
 
-def test_sorter():
```
Comment (monkey0head) [inline tests/test_metrics.py:426]: please add test for you `filter_sort` and duplicates drop

**Review by monkey0head — CHANGES_REQUESTED** [review]

File: replay/metrics/surprisal.py [hunk 3]
```diff
@@ -81,41 +76,35 @@ def _get_enriched_recommendations(
...
-                        ),
-                    ),
-                ],
-            ),
-        )
+
         recommendations = (
-            recommendations.join(self.item_weights, on="item_idx", how="left")
-            .fillna(1.0)
+            recommendations.withColumn(
```
[the remark this answers is not in the document]
Comment (mralexdmitriy) [reply replay/metrics/surprisal.py]: added separate functions for drop duplicates

**Comment by monkey0head** [comment]
please update the code from main branch

**Comment by mralexdmitriy** [comment]
code updated

File: replay/metrics/surprisal.py [hunk 4]
```diff
@@ -81,41 +76,27 @@ def _get_enriched_recommendations(
...
             .agg(
                 sf.collect_list(
                     sf.struct("relevance", "item_idx", "rec_weight")
                 ).alias("rel_id_weight")
             )
-            .withColumn("pred_rec_weight", sort_udf(sf.col("rel_id_weight")))
-            .select(
-                "user_idx",
-                sf.col("pred_rec_weight.rec_weight").alias("rec_weight"),
+            .withColumn(
```
[the remark this answers is not in the document]
Comment (mralexdmitriy) [reply replay/metrics/surprisal.py]: fixed in the last commits

File: tests/test_metrics.py [hunk 5]
```diff
@@ -423,15 +423,34 @@ def test_duplicate_recs(quality_metrics, duplicate_recs, recs, true):
...
-    assert weights == [5, 3]
+def test_filter_sort(spark, duplicate_recs):
+    recs = filter_sort(duplicate_recs)
+    gt = spark.createDataFrame(
+        data=[
+            [0, [0, 1, 2]],
+            [1, [1, 0, 4]],
+            [2, [0, 3, 2]]
+        ],
+        schema=PREDICT_SCHEMA)
```
[the remark this answers is not in the document]
Comment (mralexdmitriy) [reply tests/test_metrics.py]: fixed in the last commits

File: replay/metrics/base_metric.py [hunk 6]
```diff
@@ -102,6 +68,66 @@ def preprocess_gt(
...
+                Window.partitionBy("user_idx", "item_idx").orderBy(sf.col("relevance").desc())
+            ),
+        )
+        .where(sf.col("_num") == 1)
+        .drop("_num")
+    )
+
+
+def filter_sort(recommendations: DataFrame, *extra_columns: str) -> DataFrame:
+    """
```
Comment (monkey0head) [inline replay/metrics/base_metric.py:89]: could you please explain, why do you use the  `extra_columns` args here, not a single column name? you then define an `extra_col_type` as a type of the first extra column in the list and refer to extra_columns[0] below many times (ignore the rest of the extra columns and return the result for the first column in list only). shall we leave one `extra_column` in params?

**Review by monkey0head — CHANGES_REQUESTED** [review]

File: replay/metrics/base_metric.py (hunk 6 above)
Comment (mralexdmitriy) [reply replay/metrics/base_metric.py:89]: done

File: replay/metrics/base_metric.py [hunk 7]
```diff
@@ -102,6 +68,66 @@ def preprocess_gt(
...
+
+    recommendations = (
+        recommendations
+        .groupby("user_idx")
+        .agg(sf.collect_list(sf.struct("relevance", "item_idx", *extra_columns)).alias("pred_list"))
+        .withColumn("pred_list", sf.reverse(sf.array_sort("pred_list")))
+    )
+
+    selection = ["user_idx", sf.col("pred_list.item_idx").alias("pred")]
+    if extra_columns:
```
[the remark this answers is not in the document]
Comment (mralexdmitriy) [reply replay/metrics/base_metric.py]: done

**Review by monkey0head — APPROVED** [review]
